node/src/components/chainspec_loader.rs

```diff
@@ -231,42 +216,371 @@ impl ChainspecLoader {
+    ) -> Effects<Event> {
+        match result {
+            Ok(upgrade_result) => match upgrade_result {
+                UpgradeResult::RootNotFound
+                | UpgradeResult::KeyNotFound(_)
+                | UpgradeResult::TypeMismatch(_)
+                | UpgradeResult::Serialization(_) => {
+                    error!("failed to upgrade contract runtime: {}", upgrade_result);
+                    self.reactor_exit = Some(ReactorExit::ProcessShouldExit(ExitCode::Abort));
+                }
+                UpgradeResult::Success {
+                    post_state_hash,
+                    effect,
+                } => {
+                    info!("chainspec name {}", self.chainspec.network_config.name);
+                    info!("state root hash {}", post_state_hash);
+                    trace!(%post_state_hash, ?effect);
+                    self.reactor_exit = Some(ReactorExit::ProcessShouldContinue);
+                    self.state_root_hash = Some(post_state_hash.into());
+                }
+            },
+            Err(error) => {
+                error!("failed to upgrade contract runtime: {}", error);
+                self.reactor_exit = Some(ReactorExit::ProcessShouldExit(ExitCode::Abort));
+            }
+        }
+        Effects::new()
+    }
+
+    fn new_chainspec_info(&self) -> ChainspecInfo {
+        ChainspecInfo::new(
+            self.chainspec.network_config.name.clone(),
+            self.state_root_hash,
+            self.next_upgrade.clone(),
+        )
+    }
+
+    fn check_for_next_upgrade<REv>(&self, effect_builder: EffectBuilder<REv>) -> Effects<Event>
+    where
+        REv: From<ChainspecLoaderAnnouncement> + Send,
+    {
+        let root_dir = self.root_dir.clone();
+        let current_version = self.chainspec.protocol_config.version.clone();
+        async move {
+            let maybe_next_upgrade =
+                task::spawn_blocking(move || next_upgrade(root_dir, current_version))
+                    .await
+                    .unwrap_or_else(|error| {
+                        warn!(%error, "failed to join tokio task");
+                        None
+                    });
+            if let Some(next_upgrade) = maybe_next_upgrade {
+                effect_builder
+                    .announce_upgrade_activation_point_read(next_upgrade)
+                    .await
+            }
+        }
+        .ignore()
+    }
+
+    fn handle_got_next_upgrade(&mut self, next_upgrade: NextUpgrade) -> Effects<Event> {
+        debug!("got {}", next_upgrade);
+        if let Some(ref current_point) = self.next_upgrade {
+            if next_upgrade != *current_point {
+                info!(
+                    new_point=%next_upgrade.activation_point,
+                    %current_point,
+                    "changing upgrade activation point"
+                );
+            }
+        }
+        self.next_upgrade = Some(next_upgrade);
+        Effects::new()
+    }
```

Comment:
I read the whole PR, so I should know it at this point, but: How does the new activation point get communicated to all the components? (Probably most of them don't need to know?)

Reply:
This event is only handled by components that need to decide when to upgrade (consensus). All the components that use the chainspec will get initialized with a correct one upon startup.